Implementation of data channels on top of webrtc data channels

tdrz opened the pull request:
Implementation of #53 

Not implemented:

- [ ] Tests for DataConsumer
- [ ] SendHandler::StopSendingData() and SendHandler::StopReceivingData() are only stubs - question is if they should be implemented
- [ ] DataChannel's priority

Commit:
implementation

ibc commented:
Thanks for this. We'll take a look when possible (unfortunately very busy lately).

tdrz commented:
> Thanks for this.

Thank you guys for a great system and implementation.

>  We'll take a look when possible (unfortunately very busy lately).

No worries, there's no rush from my side. I also have to switch to something else in the mean time (that's why some things are not completely implemented).

Commit:
removed comment

Commit:
bug fixes

jmillan commented on include/DataProducer.hpp:
```diff
@@ -0,0 +1,82 @@
+    		rtc::scoped_refptr<webrtc::DataChannelInterface> webrtcDataChannel;
+    		bool closed;
+    		nlohmann::json sctpStreamParameters;
+    		nlohmann::json appData;
```
```suggestion
    		         std::string id;
			// std::string _localId;
    		        rtc::scoped_refptr<webrtc::DataChannelInterface> webrtcDataChannel;
    		        bool closed;
    		        nlohmann::json sctpStreamParameters;
    		        nlohmann::json appData;
```

jmillan commented on include/Transport.hpp:
```diff
@@ -82,18 +87,33 @@ namespace mediasoupclient
+			  	const nlohmann::json& sctpStreamParameters,
+              	const std::string& label,
+            	const std::string& protocol,
+                const nlohmann::json& appData); // not virtual so we can use only the Producer if we want to 
```
```suggestion
			  	const nlohmann::json& sctpStreamParameters,
			  	const std::string& label,
			  	const std::string& protocol,
			  	const nlohmann::json& appData); // not virtual so we can use only the Producer if we want to 
```

jmillan commented on include/Transport.hpp:
```diff
@@ -82,18 +87,33 @@ namespace mediasoupclient
+		struct DataProducerOptions : webrtc::DataChannelInit {
+    		const std::string label;
+    		const nlohmann::json appData;
+			DataProducerOptions(std::string label = "", nlohmann::json appData = nlohmann::json::object()): label(label), appData(appData){}
```
```suggestion
    		const nlohmann::json appData;
    		DataProducerOptions(std::string label = "", nlohmann::json appData = nlohmann::json::object()): label(label), appData(appData){}
```

jmillan commented on include/Transport.hpp:
```diff
@@ -160,6 +185,24 @@ namespace mediasoupclient
+			const nlohmann::json sctpStreamParameters;
+    		const std::string label;
+			const std::string protocol;
+    		const nlohmann::json appData;
```
```suggestion
			const nlohmann::json sctpStreamParameters;
			const std::string label;
			const std::string protocol;
			const nlohmann::json appData;
```

jmillan commented on include/Transport.hpp:
```diff
@@ -170,6 +213,8 @@ namespace mediasoupclient
 		  nlohmann::json* rtpParameters,
 		  const nlohmann::json& appData = nlohmann::json::object());
+		
+		DataConsumer* ConsumeData(DataConsumer::Listener* listener,	DataConsumerOptions dataConsumerOptions);
```
```suggestion
		DataConsumer* ConsumeData(DataConsumer::Listener* listener, DataConsumerOptions dataConsumerOptions);
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    // From webrtc::DataChannelObsever
+    // The data channel state have changed.
+    void DataConsumer::OnStateChange() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan replied:
Please always new line after `MS_TRACE`

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    // From webrtc::DataChannelObsever
+    // The data channel state have changed.
```
```suggestion
    // The data channel state has changed.
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+
+    //  A data buffer was successfully received.
+    void DataConsumer::OnMessage(const webrtc::DataBuffer& buffer) {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * DataConsumer id.
+     */
+    const std::string& DataConsumer::GetId() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+
+    std::string DataConsumer::GetLocalId() const {
+        MSC_TRACE();
+        return std::to_string(this->webrtcDataChannel->id());
```
```suggestion

        return std::to_string(this->webrtcDataChannel->id());
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * Associated DataProducer id.
+     */
+    const std::string& DataConsumer::GetDataProducerId() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * Whether the DataConsumer is closed.
+     */
+    const bool DataConsumer::IsClosed() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * SCTP stream parameters.
+     */
+    const json& DataConsumer::GetSctpStreamParameters() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * DataChannel readyState.
+     */
+    webrtc::DataChannelInterface::DataState DataConsumer::GetReadyState() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * DataChannel label.
+     */
+    std::string DataConsumer::GetLabel() {
+        return this->webrtcDataChannel->label();
```
```suggestion
        MSC_TRACE();

        return this->webrtcDataChannel->label();
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     */
+    std::string DataConsumer::GetProtocol()
+    {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+     * App custom data.
+     */
+    const json& DataConsumer::GetAppData() {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    void DataConsumer::Close() {
+        if (this->closed)
+            return;
+        MSC_TRACE();
```
`MS_TRACE()` is always comes first in method body.

```suggestion
        MSC_TRACE();

        if (this->closed)
            return;
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    void DataConsumer::TransportClosed() {
+        if (this->closed)
+            return;
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();

        if (this->closed)
            return;
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+        const std::string& id,
+        rtc::scoped_refptr<webrtc::DataChannelInterface> webrtcDataChannel, 
+        const json& sctpStreamParameters, 
+        const json& appData): 
```
```suggestion
		DataProducer::Listener* listener,
		const std::string& id,
		rtc::scoped_refptr<webrtc::DataChannelInterface> webrtcDataChannel, 
		const json& sctpStreamParameters, 
		const json& appData): 
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+                privateListener(privateListener), listener(listener), id(id),
+				webrtcDataChannel(webrtcDataChannel), sctpStreamParameters(sctpStreamParameters), appData(appData)
+        {
+            MSC_TRACE();
```
```suggestion
            MSC_TRACE();
            
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+    
+    const std::string& DataProducer::GetId() const
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+	std::string DataProducer::GetLocalId() const
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    bool DataProducer::IsClosed() const
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+   	const json& DataProducer::GetSctpStreamParameters() const
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    webrtc::DataChannelInterface::DataState DataProducer::GetReadyState()
+    {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+	std::string DataProducer::GetLabel()
+    {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    std::string DataProducer::GetProtocol()
+    {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    uint64_t DataProducer::GetBufferedAmount() const 
+    {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+
+    const json& DataProducer::GetAppData() const 
+	{
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+    void DataProducer::Close() 
+	{
+        MSC_TRACE();
+        if (this->closed) return;
```
```suggestion
        MSC_TRACE();
        
        if (this->closed)
                return;

```

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    void DataProducer::Send(const webrtc::DataBuffer& buffer) {
+        MSC_TRACE();
```
```suggestion
        MSC_TRACE();
        
```

jmillan commented on src/Handler.cpp:
```diff
@@ -566,6 +648,64 @@ namespace mediasoupclient
+            this->pc->SetLocalDescription(PeerConnection::SdpType::ANSWER, sdpAnswer);
+            
+			this->hasDataChannelMediaSection = true;
+		}
```
```suggestion
		if (!this->hasDataChannelMediaSection)
		{
			 this->remoteSdp->RecvSctpAssociation();
			auto sdpOffer = this->remoteSdp->GetSdp();

			MSC_DEBUG("calling pc->setRemoteDescription() [offer:%s]", sdpOffer.c_str());
			this->pc->SetRemoteDescription(PeerConnection::SdpType::OFFER, sdpOffer);
			
			webrtc::PeerConnectionInterface::RTCOfferAnswerOptions options;
			auto sdpAnswer = this->pc->CreateAnswer(options);
			
			if (!this->transportReady) {
			            auto localSdpObject = sdptransform::parse(sdpAnswer);
			            this->SetupTransport("client", localSdpObject);
			 }

			MSC_DEBUG("calling pc->setLocalDescription() [answer: %s]", sdpAnswer.c_str());
			this->pc->SetLocalDescription(PeerConnection::SdpType::ANSWER, sdpAnswer);
            
			this->hasDataChannelMediaSection = true;
		}
```

jmillan commented on src/PeerConnection.cpp:
```diff
@@ -366,6 +366,19 @@ namespace mediasoupclient
+	rtc::scoped_refptr<webrtc::DataChannelInterface> PeerConnection::CreateDataChannel(const std::string& label,
+		const webrtc::DataChannelInit* config)
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/PeerConnection.cpp:
```diff
@@ -366,6 +366,19 @@ namespace mediasoupclient
+		} else {
+			MSC_THROW_ERROR("Failed creating data channel");
+		}
+		return webrtcDataChannel;
```
```suggestion

		return webrtcDataChannel;
```

jmillan commented on src/Transport.cpp:
```diff
@@ -249,6 +250,37 @@ namespace mediasoupclient
+		auto dataProducer = new DataProducer(this, dataProducerListener, fid.get(), res.dataChannel, 
+			res.sctpStreamParameters, dataProducerOptions.appData);
+
+		this->dataProducers[dataProducer->GetId()] = dataProducer;
```
```suggestion
		this->dataProducers[dataProducer->GetId()] = dataProducer;
		
```

jmillan commented on src/Transport.cpp:
```diff
@@ -249,6 +250,37 @@ namespace mediasoupclient
+		if (!this->hasSctpParameters)
+			MSC_THROW_ERROR("SctpParameters are mandatory when using data producer listener");
+
+		SendHandler::SendDataChannel res = this->sendHandler->CreateSendDataChannel(
```
Can we rename `res` to `sendDataChannel`?

jmillan replied:
```suggestion
		// This may throw.
		SendHandler::SendDataChannel res = this->sendHandler->CreateSendDataChannel(
```

jmillan commented on src/Transport.cpp:
```diff
@@ -249,6 +250,37 @@ namespace mediasoupclient
+		SendHandler::SendDataChannel res = this->sendHandler->CreateSendDataChannel(
+			dataProducerOptions.label, dataProducerOptions);
+
+		auto fid = this->listener->OnProduceData(this, res.sctpStreamParameters, dataProducerOptions.label,
```
Can we rename `fid` to `dataChannelId`?

jmillan commented on src/Transport.cpp:
```diff
@@ -396,6 +452,32 @@ namespace mediasoupclient
+	DataConsumer* RecvTransport::ConsumeData(DataConsumer::Listener* listener, 
+		DataConsumerOptions dataConsumerOptions)
+	{
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented on src/Transport.cpp:
```diff
@@ -396,6 +452,32 @@ namespace mediasoupclient
+		else if (!this->hasSctpParameters) 
+			MSC_THROW_TYPE_ERROR("Cannot use DataChannels with this transport. SctpParameters were not set for this transport.");
+		
+		RecvHandler::RecvDataChannel res = this->recvHandler->CreateRecvDataChannel(dataConsumerOptions.label, dataConsumerOptions);
```
can we rename `res` to `CreateRecvDataChannel`?

jmillan replied:
```suggestion
		// This may throw.
		RecvHandler::RecvDataChannel res = this->recvHandler->CreateRecvDataChannel(dataConsumerOptions.label, dataConsumerOptions);
```

jmillan commented on src/Transport.cpp:
```diff
@@ -405,13 +487,20 @@ namespace mediasoupclient
 
 			consumer->TransportClosed();
 		}
+		// ... and all DataConsumers.
```
```suggestion

		// Close all DataConsumers.
```

jmillan commented on src/Transport.cpp:
```diff
@@ -265,6 +297,15 @@ namespace mediasoupclient
 			producer->TransportClosed();
 		}
+
+		// ... and all Data Producers.
```
```suggestion
		// Close all Data Producers.
```

jmillan commented on src/Transport.cpp:
```diff
@@ -396,6 +452,32 @@ namespace mediasoupclient
+			dataConsumerOptions.appData);
+
+		this->dataConsumers[dataConsumer->GetId()] = dataConsumer;
+		return dataConsumer;
```
```suggestion
		this->dataConsumers[dataConsumer->GetId()] = dataConsumer;
		
		return dataConsumer;
```

jmillan commented on src/sdp/RemoteSdp.cpp:
```diff
@@ -163,6 +163,38 @@ namespace mediasoupclient
+            "application", // kind
+            empty, // offerRtpParameters
+            "", // streamId
+			"" // trackId
```
```suggestion
            "", // streamId
           "" // trackId
```

jmillan commented on test/include/FakeTransportListener.hpp:
```diff
@@ -44,6 +44,27 @@ class FakeSendTransportListener : public mediasoupclient::SendTransport::Listene
+			  	const nlohmann::json& sctpStreamParameters,
+              	const std::string& label,
+            	const std::string& protocol,
+                const nlohmann::json& appData)
```
```suggestion
            	mediasoupclient::SendTransport* transport,
            	const nlohmann::json& sctpStreamParameters,
            	const std::string& label,
            	const std::string& protocol,
            	const nlohmann::json& appData)
```

jmillan commented on test/src/fakeParameters.cpp:
```diff
@@ -271,6 +271,8 @@ json generateTransportRemoteParameters()
 		"sctpParameters" :
 		{
 			"port"           : 5000,
+			"OS"   			 : 1024,
+			"MIS"  			 : 1024,
```
indentation

tdrz replied:
Sorry, unsure what you mean by this one.

jmillan replied:
Just replacing tabs for spaces

```diff
diff --git a/test/src/fakeParameters.cpp b/test/src/fakeParameters.cpp
index bb86e41..1089fa9 100644
--- a/test/src/fakeParameters.cpp
+++ b/test/src/fakeParameters.cpp
@@ -271,8 +271,8 @@ json generateTransportRemoteParameters()
                "sctpParameters" :
                {
                        "port"           : 5000,
-                       "OS"                     : 1024,
-                       "MIS"                    : 1024,
+                       "OS"             : 1024,
+                       "MIS"            : 1024,
                        "numStreams"     : 2048,
                        "maxMessageSize" : 2000000
                }
```

tdrz replied:
Done

jmillan reviewed (commented):
It's looking good so far :-)

tdrz commented:
Thanks, I will (hopefully) make the changes this week.

jmillan commented:
Sure @copiltembel,

You can apply my cosmetic changes within github directly if that is more comfortable for you.

tdrz committed:
Update include/DataProducer.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update include/Transport.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update include/Transport.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update include/Transport.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update include/Transport.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Transport.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Transport.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/sdp/RemoteSdp.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update test/include/FakeTransportListener.hpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Transport.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Handler.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/PeerConnection.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/PeerConnection.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Transport.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/Transport.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Commit:
applied code changes suggestions

Commit:
indentation fix

jmillan commented on include/DataConsumer.hpp:
```diff
@@ -0,0 +1,76 @@
+		class Listener {
+			public:
+				// The data channel state have changed.
+				virtual void OnStateChange(DataConsumer* dataConsumer) { };
```
Can we provide the separate `OnOpen`, `OnClose` and optionally `OnConnecting` and `OnClosing` for symmetry with mediasoup-client?

We need to get the datachannel `state` from `DataChannelObserver` and fire the corresponding listener method.

jmillan replied:
All Listener methods must be pure virtual `= 0;`

jmillan commented on include/DataConsumer.hpp:
```diff
@@ -0,0 +1,76 @@
+				// The data channel's buffered_amount has changed.
+				// virtual void OnBufferedAmountChange(uint64_t sent_data_size) {}
+				//
+				virtual void OnTransportClosed(DataConsumer* dataConsumer) { };
```
All Listener methods must be pure virtual = 0;

jmillan commented on include/DataConsumer.hpp:
```diff
@@ -0,0 +1,76 @@
+			std::string GetProtocol();
+			const nlohmann::json& GetAppData();
+			void Close();
+			void TransportClosed();
```
This method should be private. Check it [here](https://github.com/versatica/libmediasoupclient/blob/v3/include/Consumer.hpp#L60)

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+        const json& appData):
+            listener(listener), privateListener(privateListener), id(id), dataProducerId(dataProducerId),
+            webrtcDataChannel(webrtcDataChannel), sctpParameters(sctpStreamParameters), 
+            appData(appData)
```
```suggestion
        const json& appData)
        :  listener(listener), privateListener(privateListener), id(id), dataProducerId(dataProducerId),
           webrtcDataChannel(webrtcDataChannel), sctpParameters(sctpStreamParameters), 
           appData(appData)
```

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    // The data channel state have changed.
+    void DataConsumer::OnStateChange() {
+        MSC_TRACE();
+        this->listener->OnStateChange(this);
```
Here we should emit `OnOpen` or `OnClose` according to datachannel state.

jmillan commented on src/DataConsumer.cpp:
```diff
@@ -0,0 +1,127 @@
+    // The data channel's buffered_amount has changed.
+    void DataConsumer::OnBufferedAmountChange(uint64_t sent_data_size) {
+        MSC_TRACE();
+        // should not happen on consumer (I guess)
```
You're right.

```suggestion
        // Should not happen on consumer.
```

jmillan commented on include/DataProducer.hpp:
```diff
@@ -0,0 +1,82 @@
+		public:
+			virtual void OnTransportClose(DataProducer* dataProducer) {};
+			virtual void OnBufferedAmountChange(DataProducer* dataProducer, uint64_t sent_data_size) {};
+			// virtual void OnOpen(DataProducer* dataProducer) = 0;
```
`OnOpen` and `OnClose` are missing. Obtained from `DataChannelObserver.OnStateChange`

jmillan commented on include/Handler.hpp:
```diff
@@ -10,6 +10,8 @@
 #include <api/rtp_sender_interface.h>      // webrtc::RtpSenderInterface
 #include <api/rtp_transceiver_interface.h> // webrtc::RtpTransceiverInterface
+#include <third_party/abseil-cpp/absl/types/optional.h>
+
```
```suggestion
```

jmillan commented on include/Handler.hpp:
```diff
@@ -10,6 +10,8 @@
 #include <api/rtp_receiver_interface.h>    // webrtc::RtpReceiverInterface
 #include <api/rtp_sender_interface.h>      // webrtc::RtpSenderInterface
 #include <api/rtp_transceiver_interface.h> // webrtc::RtpTransceiverInterface
+#include <third_party/abseil-cpp/absl/types/optional.h>
```
where is this used?

tdrz replied:
Leftovers from tests. Thanks.

jmillan commented on include/Transport.hpp:
```diff
@@ -170,6 +213,8 @@ namespace mediasoupclient
 		  nlohmann::json* rtpParameters,
 		  const nlohmann::json& appData = nlohmann::json::object());
+		
+		DataConsumer* ConsumeData(DataConsumer::Listener* listener,	DataConsumerOptions dataConsumerOptions);
```
In libmediasoupclient we don't use `ProducerOptions` `ConsumerOptions` but the constructor takes the needed arguments separately: dataConsumerId, dataProducerId, sctpStreamParameters, label, etc.

Lets build the `DataChannelInit` instance internally on constructor with the given arguments. There is no need to expose it to the user in this case.

tdrz replied:
OK, done. But I guess you mean the ProduceData()/ConsumeData() take the needed arguments and not the constructors, right?

jmillan commented on include/Transport.hpp:
```diff
@@ -120,13 +140,17 @@ namespace mediasoupclient
 		  const nlohmann::json* codecOptions,
 		  const nlohmann::json& appData = nlohmann::json::object());
 
+		DataProducer* ProduceData(DataProducer::Listener* listener,
```
In libmediasoupclient we don't use `ProducerOptions` `ConsumerOptions` but the constructor takes the needed arguments separately.

Lets build the `DataChannelInit` instance internally on constructor with the given arguments. There is no need to expose it to the user in this case.

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+	// The data channel state have changed.
+	void DataProducer::OnStateChange(){
+		MSC_TRACE();
+		// TODO
```
Yep, lets emit here `OnOpen` or `OnClose` accordingly.

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+  	
+	//  A data buffer was successfully received.
+  	void DataProducer::OnMessage(const webrtc::DataBuffer& buffer) {
+		//MSC_WARN("DataProducer::OnMessage() | dataChannel 'message' event in a DataProducer, message discarded");
```
We can just add a comment indicating that this will never happen and keep the body empty.

jmillan commented on src/DataProducer.cpp:
```diff
@@ -0,0 +1,121 @@
+  
+  	// The data channel's buffered_amount has changed.
+  	void DataProducer::OnBufferedAmountChange(uint64_t sent_data_size) {
+		MSC_TRACE();
```
```suggestion
		MSC_TRACE();
		
```

jmillan commented:
@copiltembel, please run `gulp format` from console in order to get the code properly formatted.

you need to install `gulp` if not already installed: `npm install -g gulp-cli`

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataConsumer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

tdrz committed:
Update src/DataProducer.cpp

Co-authored-by: José Luis Millán <jmillan@aliax.net>

Commit:
partial changes

Commit:
more changes

Commit:
test adapted to new methods

Commit:
Merge branch 'ChangesAfterCodeReview2' into v3-DataChannelsImpl

tdrz commented:
I get an "Error: Cannot find module 'gulp-util'" when running gulp format. Tried a couple of things to fix it but didn't help. Sorry!

jmillan commented:
@copiltembel, I'll do it if you give me access to the branch.

jmillan committed:
gulp format

jmillan commented:
done :-)

tdrz commented:
@jmillan Any next steps here?

jmillan commented:
> Any next steps here?

1. Update please the PR description so it reflects what's missing at this moment.
2. Once it's ready, we'll prepare mediasoup-demo so we can use mediasoup-broadcaster-demo with datachannels for testing.

tdrz commented:
Sounds good, will hopefully have time next week to revisit everything.

tdrz committed:
Merge pull request #1 from versatica/v3

Merge newest code into local branch

tdrz commented:
@jmillan 

Looking at SendHandler::CreateSendDataChannel(...):

1.  I believe I should call "ortc::validateSctpStreamParameters(sctpStreamParameters);" before creating the webrtcDataChannel.
2. Because of 1 I would move

> rtc::scoped_refptr<webrtc::DataChannelInterface> webrtcDataChannel = this->pc->CreateDataChannel(label, &dataChannelInit);

in the SendDataChannel constructor.

3. Also because of 1, I would change the signature of

> CreateSendDataChannel(const std::string& label, webrtc::DataChannelInit dataChannelInit)

to remove the webrtc::DataChannelInit and have plain params (reliable, ordered etc.).

4. Same for RecvDataChannel

**What do you think?**

jmillan commented:
Indeed, I see you've created `Handler::CreateSendDataChannel()` whereas in mediasoup-client the method is called `Handler::SendDataChannel()`.

Lets mimic mediasoup-client as both libraries must behave exactly the same way.

Commit:
rename createsend/recvdatachannel

tdrz commented:
@jmillan 

About the ortc.validateSctpStreamParameters(sctpStreamParameters) in **mediasoup-client**:

Probably I am misunderstading, but ortc.validateSctpStreamParameters(...) is called _after_ the handler.sendDataChannel(...) is called. Wouldn't it make sense to validate/fill in missing fields _before_ actually creating the SendDataChannel?

From [here](https://github.com/versatica/mediasoup-client/blob/8c605689afdf1d924abda053243f950cc985b327/src/Transport.ts#L668).

tdrz commented:
About DataChannel priority:

IIUC: datachannel priority is not supported by any browser: https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel

Also skimming through webrtc's (m84) code, I couldn't find this parameter for DataChannels. Most notably, this is not present on _webrtc::DataChannelInit_ struct.

tdrz committed:
Merge pull request #2 from copiltembel/v3-DataChannelsImpl-Refactorings

V3 data channels impl refactorings

Commit:
Revert previous commit; make it build

Commit:
leftovers

jmillan commented:
> Probably I am misunderstading, but ortc.validateSctpStreamParameters(...) is called after the handler.sendDataChannel(...) is called. Wouldn't it make sense to validate/fill in missing fields before actually creating the SendDataChannel?

Feel free to call it before.

jmillan commented:
> About DataChannel priority:

> IIUC: datachannel priority is not supported by any browser: https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel

> Also skimming through webrtc's (m84) code, I couldn't find this parameter for DataChannels. Most notably, this is not present on webrtc::DataChannelInit struct.

Correct, please don't include it.

Commit:
validate sctp stream parameters before creating the data channel

Commit:
removed data channel priority as it is not supported by newer versions of webrtc

Commit:
refactorings

tdrz committed:
Merge pull request #3 from copiltembel/v3-DataChannelsImpl-Retake

V3 data channels impl retake

tdrz commented:
@jmillan 

I've updated the branch with the discussed changes.

I have also implemented a minimal demo:

mediasoup-demo: https://github.com/copiltembel/mediasoup-demo/tree/AddLibmediasoupDataChannelDemo

mediasoup-broadcaster-demo: https://github.com/copiltembel/mediasoup-broadcaster-demo-datachannels - here I've added libmediasoup as a submodule, so if you checkout the code, make sure to do **git clone ...** and then **git submodule init** and **git submodule update** to fetch all the data.

The demo is a simple loopback dataproducer->dataconsumer in mediasoup-broadcaster-demo. It would be nicer to broadcast the message to all of the peers, but I don't have the time to implement it now. Also related to this, it would be nice if protoo would have c++ bindings; it would make the above easier.

jmillan commented:
Thanks @copiltembel,

We may need some days to check it, but it's not forgotten.

tdrz commented:
I've added the "disableTrackOnPause" and "zeroRtpOnPause" options (and logic) on my private branch. I could add the changes to this PR if you find it useful. Let me know.

jmillan committed:
Merge remote-tracking branch 'origin/v3' into v3-DataChannelsImpl

jmillan committed:
Data[Consumer|Producer]: Add missing MSC_CLASS definition

jmillan committed:
test: fix DataChannel tests

jmillan committed:
update .clang-tidy

jmillan committed:
cosmetic

jmillan commented:
> I've added the "disableTrackOnPause" and "zeroRtpOnPause" options (and logic) on my private branch. I could add the changes to this PR if you find it useful. Let me know.

In another PR please.

jmillan commented:
> mediasoup-broadcaster-demo: https://github.com/copiltembel/mediasoup-broadcaster-demo-datachannels - here I've added libmediasoup as a submodule, so if you checkout the code, make sure to do git clone ... and then git submodule init and git submodule update to fetch all the data.

Can you please make a PR to [mediasoup-broadcaster-demo](https://github.com/versatica/mediasoup-broadcaster-demo) adding datachannels, so we just maintain 1 demo for this purposes?

jmillan committed:
remove commented line

jmillan committed:
format

jmillan committed:
Make Data[Producer|Consumer] constructors private

jmillan committed:
remove commented lines

jmillan committed:
cosmetic

jmillan committed:
Handler: unify recv and send DataChannel struct and cleanup

jmillan committed:
DataConsumer: add log for unknown data channel state

jmillan committed:
remove comment

jmillan committed:
DataChannel naming

jmillan committed:
more DataChannel naming

jmillan committed:
more on DataChannel naming

jmillan committed:
format

jmillan committed:
DataConsumer: remove unnecessary include statements

jmillan committed:
DataProducer: remove unnecessary include statements

jmillan committed:
cosmetic

jmillan committed:
cosmetic

jmillan committed:
data channel code cleanup

jmillan committed:
Handler: consistency with media and data send|recv

jmillan committed:
cosmetic

jmillan committed:
Data[Consumer|Producer] reorder methods

jmillan committed:
make 'protocol' optional in RecvTransport::ConsumeData

tdrz commented:
> > mediasoup-broadcaster-demo: https://github.com/copiltembel/mediasoup-broadcaster-demo-datachannels - here I've added libmediasoup as a submodule, so if you checkout the code, make sure to do git clone ... and then git submodule init and git submodule update to fetch all the data.
> 
> Can you please make a PR to [mediasoup-broadcaster-demo](https://github.com/versatica/mediasoup-broadcaster-demo) adding datachannels, so we just maintain 1 demo for this purposes?

Sure, but it would be easier for me if I do it after you complete this PR. Then I can just link directly the v3 libmediasoupclient branch to the mediasoup-broadcaster-demo.

jmillan commented:
> Sure, but it would be easier for me if I do it after you complete this PR. Then I can just link directly the v3 libmediasoupclient branch to the mediasoup-broadcaster-demo.

As for now this PR is complete, but we need to test it within mediasoup-broadcaster-demo ideally.

If you could make a PR for mediasoup-broadcaster-demo and point *this* libmediasoupclient branch in the `scripts/get_deps.sh` so we can test before merging to v3, would be great.

tdrz commented:
OK, will do.

tdrz commented:
@jmillan 

Funny enough I can't build mediasoup-broadcaster-demo anymore:

/usr/bin/ld: [...]/webrtc/src/out/m84/obj/libwebrtc.a(latebindingsymboltable_linux.o): undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/libdl.so: error adding symbols: DSO missing from command line

Did you encounter this problem before?

tdrz commented:
I fixed it temporarily for me by adding libwebrtc to the target_link_libraries:

```
	# Public (interface) dependencies.
target_link_libraries(${PROJECT_NAME} PUBLIC
	${CPR_LIBRARIES}
	mediasoupclient
	webrtc_broadcaster
	${LIBWEBRTC_BINARY_PATH}/libwebrtc${CMAKE_STATIC_LIBRARY_SUFFIX}
)
```

tdrz commented:
@ibc Done. Please have a look at 

https://github.com/versatica/mediasoup-broadcaster-demo/pull/10 and
https://github.com/versatica/mediasoup-demo/pull/101

I couldn't solve the [problem mentioned a couple of comments above](https://github.com/versatica/libmediasoupclient/pull/77#issuecomment-647045491) in any other way than targeting the libwebrtc directly in the mediasoup-broadcaster-demo.

jmillan commented:
We'll check it as soon as possible.

tdrz commented:
I just discovered a bug, please wait with the review.

tdrz commented:
Fixed.

jmillan committed:
Merge branch 'v3' into HEAD

jmillan commented:
merged

jmillan commented:
Thanks @copitembel!
